### PR TITLE
SearchBackPressure Service Node/Cluster RCA

### PR DESCRIPTION
**Pull request by CoderJeffrey**
**Is your feature request related to a problem? Please provide an existing Issue # , or describe.**
After Reader Component of RCA digest Search Back Pressure service related stats and populate the stats in sqlitedb, we need Cluster/Node level RCA to analyze the stats and send node/cluster level Resource Flow Units to deciders. Then, deciders can use the resource flow units to suggest actions for downstream classes. 

**Describe the solution you are proposing**
Create Cluster/Node Level RCA to generate Resource FlowUnit by comparing stats of SearchBackPressure Service and threshold, and added corresponding UTs.

Search Back Pressure RCA analyze the stats of `Searchbp_Stats Table` in sqlitedb from previous PR of Reader Component for RCA (https://github.com/opensearch-project/performance-analyzer-rca/pull/427). Example stats (e.g heap_cancellation_count with threshold, generate healthy/unhealthy resource flow units) are used to determine the health status of a node.

**Describe alternatives you've considered**
N/A

**Additional context**
The logic to decide whether the node level rca healthiness is not decided yet, we will talk with other teams (e.g. SearchBackPressure team) to discuss the final threshold for RCA, and the logic will likely change in near future.

### Check List
- [X] New functionality includes testing.
- [X] All tests pass
- [X] New functionality has been documented.
- [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -371,7 +371,7 @@ dependencies {
     implementation("org.mockito:mockito-core") {
         version {
             strictly "2.23.0"
-        }
+        }
```
Comment (ansjcy) [inline build.gradle:374]: nit: avoid trailing space

File: src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    public static final int DEFAULT_MIN_HEAP_DECREASE_THRESHOLD = 80;
+    private Integer minHeapDecreasePercentageThreshold;
+
+    // cancellationCount due to heap is less than 30% of all task cancellations in shard level
+    public static final int DEFAULT_SHARD_MIN_HEAP_CANCELLATION_THRESHOLD = 30;
+    private Integer minShardHeapCancellationPercentageThreshold;
+
+    // cancellationCount due to heap is less than 30% of all task cancellations in task level
+    public static final int DEFAULT_TASK_MIN_HEAP_CANCELLATION_THRESHOLD = 30;
+    private Integer minTaskHeapCancellationPercentageThreshold;
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java:50]: will the values of these thresholds be modified during runtime? If not we can make them `final` as well.

File: src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    public Integer getMinShardHeapCancellationPercentageThreshold() {
+        return minShardHeapCancellationPercentageThreshold;
+    }
+
+    public Integer getMinTaskHeapCancellationPercentageThreshold() {
+        return minTaskHeapCancellationPercentageThreshold;
+    }
+
+    // name for the configuration field
+    public static class RCA_CONF_KEY_CONSTANTS {
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java:125]: It might be better to use `enum` instead of `public static class` here to define a list of related constants.
In general, enums offer better readability, type safety, and if we need to iterating over those constants we can simply use the `values()` method. This is more straightforward than manually iterating over constants in a public static final class.
you can do something like:
```
    public enum RCA_CONF_KEY_CONSTANTS {
        MAX_HEAP_USAGE_INCREASE_FIELD("max-heap-usage-increase"),
        MAX_SHARD_HEAP_CANCELLATION_PERCENTAGE_FIELD("max-shard-heap-cancellation-percentage");
        // ...

        private final String value;
        RCA_CONF_KEY_CONSTANTS(final String value) {
            this.value = value;
        }
        @Override
        public String toString() {
            return value;
        }
    }
```

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/OpenSearchAnalysisGraph.java [hunk 4]
```diff
@@ -433,6 +443,28 @@ public void construct() {
...
+        searchBackPressureClusterRCA.addTag(
+                RcaConsts.RcaTagConstants.TAG_LOCUS,
+                RcaConsts.RcaTagConstants.LOCUS_CLUSTER_MANAGER_NODE);
+        searchBackPressureClusterRCA.addAllUpstreams(
+                Collections.singletonList(searchBackPressureRCA));
+        searchBackPressureClusterRCA.addTag(
+                RcaConsts.RcaTagConstants.TAG_AGGREGATE_UPSTREAM,
+                RcaConsts.RcaTagConstants.LOCUS_DATA_NODE);
+
+        // To Do SearchBackPressure Decider
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/store/OpenSearchAnalysisGraph.java:466]: nit: use `TODO` instead of `To Do` for get better IDE support

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/OldGenRca.java [hunk 5]
```diff
@@ -250,4 +250,61 @@ public double readMin() {
             return Double.NaN;
         }
     }
+
+    /**
+     * Sliding window to check the max/min olg gen usage within a given time frame
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/OldGenRca.java:255]: nit: olg -> old

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/OldGenRca.java [hunk 6]
```diff
@@ -250,4 +250,61 @@ public double readMin() {
...
+                int SLIDING_WINDOW_SIZE_IN_TIMESTAMP,
+                TimeUnit timeUnit,
+                boolean isMinSlidingWindow) {
+            super(SLIDING_WINDOW_SIZE_IN_TIMESTAMP, timeUnit);
+            this.isMinSlidingWindow = isMinSlidingWindow;
+        }
+
+        @Override
+        public void next(SlidingWindowData e) {
+            boolean pollFirstCondition;
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/OldGenRca.java:274]: This variable `pollFirstCondition` is not used. We can remove it

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/searchbackpressure/SearchBackPressureRCA.java [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+        this.shardJVMCancellationSlidingWindow =
+                new SlidingWindow<>(SLIDING_WINDOW_SIZE_IN_MINS, TimeUnit.MINUTES);
+        this.taskJVMCancellationSlidingWindow =
+                new SlidingWindow<>(SLIDING_WINDOW_SIZE_IN_MINS, TimeUnit.MINUTES);
+
+        LOG.info("SearchBackPressureRCA initialized");
+    }
+
+    /*
+     * operate() is used for local build
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/searchbackpressure/SearchBackPressureRCA.java:146]:
```
operate() is used for local build
```
is this related to this function?

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/searchbackpressure/SearchBackPressureRCA.java [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+                searchBackPressureRCAMetric.getMaxHeap(),
+                searchBackPressureRCAMetric.getHeapUsagePercent(),
+                searchBackPressureRCAMetric.getSearchbpShardCancellationCount(),
+                searchBackPressureRCAMetric.getSearchbpTaskCancellationCount(),
+                searchBackPressureRCAMetric.getSearchbpJVMShardCancellationCount(),
+                searchBackPressureRCAMetric.getSearchbpJVMTaskCancellationCount());
+
+        updateAllSlidingWindows(searchBackPressureRCAMetric, currentTimeMillis);
+
+        LOG.info("SearchBackPressureRCA currentIterationNumber is {}", currentIterationNumber);
```
Comment (ansjcy) [inline src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/searchbackpressure/SearchBackPressureRCA.java:190]: we can change some of the `LOG.info` messages to `LOG.debug`. `LOG.info` should only contains essential information to store in the service log.

File: build.gradle (hunk 1 above)
Comment (CoderJeffrey) [reply build.gradle]: removed

File: src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java (hunk 2 above)
Comment (CoderJeffrey) [reply src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java:50]: Yes, these values like **minTaskHeapCancellationPercentageThreshold** will read values from config file in runtime thus we cannot mark them as **final**

`        
minHeapDecreasePercentageThreshold =
                conf.readRcaConfig(
                        CONFIG_NAME,
                        RCA_CONF_KEY_CONSTANTS.MAX_HEAP_USAGE_DECREASE_FIELD,
                        DEFAULT_MIN_HEAP_DECREASE_THRESHOLD,
                        (s) -> s >= 0 && s <= 100,
                        Integer.class);
`

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/OpenSearchAnalysisGraph.java (hunk 4 above)
Comment (CoderJeffrey) [reply src/main/java/org/opensearch/performanceanalyzer/rca/store/OpenSearchAnalysisGraph.java]: Fixed

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/OldGenRca.java (hunk 6 above)
Comment (CoderJeffrey) [reply src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/OldGenRca.java]: removed

File: src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/searchbackpressure/SearchBackPressureRCA.java (hunk 8 above)
Comment (CoderJeffrey) [reply src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/searchbackpressure/SearchBackPressureRCA.java]: Change unnecessary log message level to debug

File: src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java (hunk 3 above)
Comment (CoderJeffrey) [reply src/main/java/org/opensearch/performanceanalyzer/rca/configs/SearchBackPressureRcaConfig.java]: Thank you so much for the suggestion. I have switched the design from static class to enum

**Review by khushbr — APPROVED** [review]

**Review by ansjcy — APPROVED** [review]
